### PR TITLE
Switch from vanilla KCL processor to Akka streams

### PR DESCRIPTION
**Pull request by sihil**
## What does this change?
This switches Thrall from using the vanilla KCL library to consuming the Kinesis stream via an akka stream.

This runs locally and there is evidence that it correcty processes and checkpoints events.

From here it should be a hop, skip and a jump to add a second incoming kinesis stream with priority merging using a `MergePrioritised` graph shape.

## How can success be measured?
This should work in exactly the same way as the existing code.

## Who should look at this?

Pre-rollout
-----------

 - [x] CFN update must be applied for new permission: https://github.com/guardian/editorial-tools-platform/pull/412

## Tested?
- [X] locally
- [x] on TEST

### CODE REVIEW

File: thrall/app/lib/kinesis/ThrallEventConsumer.scala [hunk 1]
```diff
@@ -65,13 +66,14 @@ class ThrallEventConsumer(es: ElasticSearch,
     }
   }
 
-  private def parseRecord(r: Record):Option[UpdateMessage] = {
+  def parseRecord(r: Record):Option[UpdateMessage] = parseRecord(r.getData.array, r.getApproximateArrivalTimestamp.toInstant)
+
+  def parseRecord(r: Array[Byte], timestamp: Instant):Option[UpdateMessage] = {
```
Comment (jonathonherbert) [inline thrall/app/lib/kinesis/ThrallEventConsumer.scala]: Is there a reason this is overloaded in this way? I'm reading this as sugar for two val assignments at the start of the second declaration of `parseRecord`, perhaps to clarify the way the `Record` is used in the function?

File: thrall/app/controllers/HealthCheck.scala [hunk 2]
```diff
@@ -4,18 +4,17 @@ import com.gu.mediaservice.lib.argo.ArgoHelpers
...
 import lib.elasticsearch._
-import lib.kinesis.ThrallMessageConsumer
 import play.api.Logger
 import play.api.mvc._
 
-import scala.concurrent.ExecutionContext
+import scala.concurrent.{ExecutionContext, Future}
 
-class HealthCheck(elasticsearch: ElasticSearch, messageConsumer: ThrallMessageConsumer, config: ThrallConfig, override val controllerComponents: ControllerComponents)(implicit override val ec: ExecutionContext)
+class HealthCheck(elasticsearch: ElasticSearch, streamRunning: Future[_], config: ThrallConfig, override val controllerComponents: ControllerComponents)(implicit override val ec: ExecutionContext)
```
Comment (akash1810) [inline thrall/app/controllers/HealthCheck.scala]: This type signature looks really odd 🤔

File: thrall/app/lib/ThrallStreamProcessor.scala [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      from = config.from
+    ).withKinesisEndpoint(thrallKinesisEndpoint)
+     .withDynamoDBEndpoint(thrallKinesisDynamoEndpoint)
+  }
+
+  val kinesisSource: Source[KinesisRecord, Future[Done]] = KinesisSource(kinesisConfig)
+
+  implicit val dispatcher = actorSystem.getDispatcher
+
+  def run(): Future[Done] = {
```
Comment (akash1810) [inline thrall/app/lib/ThrallStreamProcessor.scala]: Not quite sure I understand how we're returning a `Future[Done]` here.

File: thrall/app/lib/kinesis/ThrallEventConsumer.scala (hunk 1 above)
Comment (sihil) [reply thrall/app/lib/kinesis/ThrallEventConsumer.scala]: This was overloaded as I had originally preserved the code of both ways (direct KCL and KCL via Akka streams) of consuming the Kinesis stream and they produced different types. As such one was a shim for the other without muddying the calling code.

However I removed the dual code towards the end of my work but didn't clean up the methods in this class.

File: thrall/app/lib/ThrallStreamProcessor.scala (hunk 3 above)
Comment (sihil) [reply thrall/app/lib/ThrallStreamProcessor.scala]: `Future[Done]` is essentially `Future[Unit]`. It is returned by akka streams - in this case from the `runForEach`.

File: thrall/app/controllers/HealthCheck.scala (hunk 2 above)
Comment (sihil) [reply thrall/app/controllers/HealthCheck.scala]: Is `=> Boolean` better? Leave the implementation to the wiring?

**Pull request merged**